cloudconfig/containerinit/container_userdata.go

```diff
@@ -380,3 +385,103 @@ else
 fi`[1:],
 		oldInterfacesFile, newInterfacesFile)
 }
+
+func populateNetworkInterfaces(networkFile string) string {
+	s := `
+if [ -f /usr/bin/python ]; then
+    python %s.py --interfaces_file %s
+else
+    python3 %s.py --interfaces_file %s
+fi
+`
+	return fmt.Sprintf(s, networkFile, networkFile, networkFile, networkFile)
+}
+
+func PopulateNetworkInterfacesScript() string {
+	s := `import subprocess, re, argparse, os, time
+from string import Formatter
+INTERFACES_FILE="/etc/network/interfaces"
+IP_LINE = re.compile(r"^\d: (.*?):")
+IP_HWADDR = re.compile(r".*link/ether ((\w{2}|:){11})")
+COMMAND = "ip -oneline link"
+RETRIES = 3
+WAIT = 5
+
+# Python3 vs Python2
+try:
+    strdecode = str.decode
+except AttributeError:
+    strdecode = str
+
+def ip_parse(ip_output):
+    """parses the output of the ip command
+    and returns a hwaddr->nic-name dict"""
+    devices = dict()
+    print("Parsing ip command output %s" % ip_output)
+    for ip_line in ip_output:
+        ip_line_str = strdecode(ip_line, 'utf-8')
+        match = IP_LINE.match(ip_line_str)
+        if match is None:
+            continue
+        nic_name = match.group(1)
+        match = IP_HWADDR.match(ip_line_str)
+        if match is None:
+            continue
+        nic_hwaddr = match.group(1)
+        devices[nic_hwaddr]=nic_name
+    print("Found the following devices: %s" % str(devices))
+    return devices
+
+def replace_ethernets(interfaces_file, devices, fail_on_missing):
+    """check if the contents of interfaces_file contain template
+    keys corresponding to hwaddresses and replace them with
+    the proper device name"""
+    with open(interfaces_file, "r") as intf_file:
+        interfaces = intf_file.read()
+
+    formatter = Formatter()
+    hwaddrs = [v[1] for v in formatter.parse(interfaces) if v[1]]
+    print("Found the following hwaddrs: %s" % str(hwaddrs))
+    device_replacements = dict()
+    for hwaddr in hwaddrs:
+        hwaddr_clean = hwaddr[3:].replace("_", ":")
+        if devices.get(hwaddr_clean, None):
+            device_replacements[hwaddr] = devices[hwaddr_clean]
+        else:
+            if fail_on_missing:
+                print("Can't find device with MAC %s, will retry" % hwaddr_clean)
+                return False
+            else:
+                print("WARNING: Can't find device with MAC %s when expected" % hwaddr_clean)
+                device_replacements[hwaddr] = hwaddr
+    formatted = interfaces.format(**device_replacements)
+    print("Used the values in: %s\nto fix the interfaces file:\n%s\ninto\n%s" %
+           (str(device_replacements), str(interfaces), str(formatted)))
+
+    with open(interfaces_file + ".tmp", "w") as intf_file:
+        intf_file.write(formatted)
+
+    os.rename(interfaces_file, interfaces_file + ".bak")
```

Comment:
One thing I noticed in testing, we're writing *something* to a wrong location, because /etc/network/interface is *not* the one that ships with ubuntu/cloud-init.
The .bak file that is saved is the content that we wrote that we are expanding the template. We should make sure that we write:
 a) /etc/network/interfaces.tmpl  (the template {eth_ab_cd_ef_gh} form
 b) /etc/network/interfaces.bak a backup of what was originally written as /e/n/i
 c) /etc/network/interfaces.py
 d) /etc/network/interfaces the final rendered script